Enhancing the InfoEndpoint so it is possible to provide additional information besides git.properties and application.properties

mat013 opened the pull request:
The current implementation of InfoEndpoint can only retrieve data from application property files and git property file. 

I would like to add an enhancement where one can add additional properties from other sources than git.properties file and application.properties.

It is inspired by the HealthEndpoint mechanism, where one can register a bean with a specific interface and the HealthEndpoint will pick up the bean and extract data from that bean.

In the case of the InfoEndpoint one has to implement the org.springframework.boot.actuate.info.InfoProvider and register it within the application context

The constructor for the InfoEndpoint class has been altered to get InfoProviders, which it should extract from (again same principle as the HealthPoint)

snicoll commented on spring-boot-actuator/src/main/java/org/springframework/boot/actuate/info/InfoProvider.java:
```diff
@@ -0,0 +1,17 @@
+	/**
+	 * @return a collection of information
+	 */
+	Map<String, Object> provide();
```
I'd prefer a more fine-grained API _à la_ `Health`

mat013 replied:
Hi Stephane

I can do that but I think I will break the interface when I change the
interface to return another type than Map no?

Or do you have any suggestions?

br

Meang


On Tue, Jul 14, 2015 at 9:21 PM, Stéphane Nicoll <notifications@github.com>
wrote:

> In
> spring-boot-actuator/src/main/java/org/springframework/boot/actuate/info/InfoProvider.java
> <https://github.com/spring-projects/spring-boot/pull/3492#discussion_r34607930>
> :
>
> > @@ -0,0 +1,17 @@
> > +package org.springframework.boot.actuate.info;
> > +
> > +import java.util.Map;
> > +
> > +/**
> > + * information provider for the info endpoint
> > + *
> > + * @author Meang Akira Tanaka
> > + */
> > +public interface InfoProvider {
> > +
> > +	/**
> > +	 * @return a collection of information
> > +	 */
> > +	Map<String, Object> provide();
>
> I'd prefer a more fine-grained API *à la* Health
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-projects/spring-boot/pull/3492/files#r34607930>
> .
>

snicoll replied:
Break what interface? You just created it :)

If you return a strongly-typed object, it gives you the ability to evolve said object in the future without breaking anything. In general returning simple types like map from an interface is a bad idea. 

I am not too sure what to do at this point but it feels to me that the default implementation of `InfoProvider` should read the `info.*` keys from the context.

mat013 replied:
Hi Stephane

I agree that it would be better to return it's own type rather than a basic
type as map.

however i returned the map because the old implementation of the invoke
method at InfoeEndpoint returns a map. So my question was asked if it would
be OK to break the signature of the original method?

> I am not too sure what to do at this point but it feels to me that the
default implementation of InfoProvidershould read the info.* keys from the
context.

I think the infoendpoint is already doing that. Should that be moved to a
default implementation?

Or is the name infoprovider correct?

Br

Meang



On Wed, Jul 15, 2015, 10:06 Stéphane Nicoll <notifications@github.com>
wrote:

> In
> spring-boot-actuator/src/main/java/org/springframework/boot/actuate/info/InfoProvider.java
> <https://github.com/spring-projects/spring-boot/pull/3492#discussion_r34654867>
> :
>
> > @@ -0,0 +1,17 @@
> > +package org.springframework.boot.actuate.info;
> > +
> > +import java.util.Map;
> > +
> > +/**
> > + * information provider for the info endpoint
> > + *
> > + * @author Meang Akira Tanaka
> > + */
> > +public interface InfoProvider {
> > +
> > +	/**
> > +	 * @return a collection of information
> > +	 */
> > +	Map<String, Object> provide();
>
> Break what interface? You just created it :)
>
> If you return a strongly-typed object, it gives you the ability to evolve
> said object in the future without breaking anything. In general returning
> simple types like map from an interface is a bad idea.
>
> I am not too sure what to do at this point but it feels to me that the
> default implementation of InfoProvider should read the info.* keys from
> the context.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-projects/spring-boot/pull/3492/files#r34654867>
> .
>

snicoll replied:
If we accept this PR, we'll have to break `InfoEndpoint`, yes.

I also think that the business of reading `info.*` should move to a default implementation of `InfoProvider`. You may want to review the health support as it has quite a lot of similarities with what you're trying to do. Thanks!

mat013 replied:
Ok i will do that

Have a nice day :-)

On Wed, Jul 15, 2015, 10:25 Stéphane Nicoll <notifications@github.com>
wrote:

> In
> spring-boot-actuator/src/main/java/org/springframework/boot/actuate/info/InfoProvider.java
> <https://github.com/spring-projects/spring-boot/pull/3492#discussion_r34656295>
> :
>
> > @@ -0,0 +1,17 @@
> > +package org.springframework.boot.actuate.info;
> > +
> > +import java.util.Map;
> > +
> > +/**
> > + * information provider for the info endpoint
> > + *
> > + * @author Meang Akira Tanaka
> > + */
> > +public interface InfoProvider {
> > +
> > +	/**
> > +	 * @return a collection of information
> > +	 */
> > +	Map<String, Object> provide();
>
> If we accept this PR, we'll have to break InfoEndpoint, yes.
>
> I also think that the business of reading info.* should move to a default
> implementation of InfoProvider. You may want to review the health support
> as it has quite a lot of similarities with what you're trying to do. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-projects/spring-boot/pull/3492/files#r34656295>
> .
>

mat013 commented:
Hi 

I've changed the return type as Stephane suggested to a more descriptive type, which has caused a breaking change to the InfoEndpoint.

In addition he suggested to implement an info provider which I guess should replace the behaviour of the one implemented in the EndpointAutoConfiguration, so I have refactored the git info and the environment info out of EndpointAutoConfiguration and implemented two info providers ScmGitPropertiesInfoProvider and EnvironmentInfoProvider instead. in order to register these two I looked at how the health endpoint did, and i implemented the InfoProviderAutoConfiguration in a similar fashion, but I am by no means an expert on implementing an AutoConfiguration, so please let me know if that is the right way to do so.

The thing I did to create an AutoConfiguration is as follows:

1) I've implemented InfoProviderAutoConfiguration
2) added an entry in the spring.factories

I am unsure if I need to provide with additional documentation but let me know if I need so.

br

Meang

jceloi commented:
Hello there,

This pull request is really interesting, I was wondering how to extend the properties and was surprised that even if the InfoEndpoint points toward extension with the getAdditionalInfo, you have to copy / rewrite the EndpointAutoConfiguration info properties factory if you intend to extend the InfoEndpoint, which seems a bit sad.

Any news regarding a merge ?

Thanks !

snicoll commented:
I've been quite busy on other things the last few days but I intend to give it another look in the coming week. Thanks.

mat013 commented:
Hi Stephan. 

Any news on whether this PR can be merged or if something needs to be changed? 

br Meang

snicoll commented:
I've been carried away by other stuff, sorry about that. I think it's fine but we have other related issues (especially related to GitInfo) that I'd like to tackle alongside this one and the 1.3 window is pretty much closed. I'd like to revisit this for 1.4

mat013 commented:
What kind of issues... maybe I can incorporate a fix to that as well...

br

Meang

On Wed, Aug 26, 2015 at 12:01 PM, Stéphane Nicoll <notifications@github.com>
wrote:

> I've been carried away by other stuff, sorry about that. I think it's fine
> but we have other related issues (especially related to GitInfo) that I'd
> like to tackle alongside this one and the 1.3 window is pretty much closed.
> I'd like to revisit this for 1.4
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-projects/spring-boot/pull/3492#issuecomment-134930927>
> .
>

snicoll commented:
More of a design issue, really. I'd like to incorporate that change as a bigger change and for that I need time to think on the big picture.

cemo commented:
duplicates #2559

snicoll commented:
@cemo not really, it's a pull request with a proposal for #2559

cemo commented:
For those who needs this, I have done something like this which is working for us currently. 

https://gist.github.com/cemo/e38a2c20bada2e6b0901

snicoll commented:
Here is an exemple of a custom `InfoContributor` that exposes all the content of `git.properties`

```java
@Bean
public InfoContributor gitFullInfoContributor() throws IOException, BindException {
	Resource resource = new ClassPathResource("git.properties");
	Map<String, Object> content = new LinkedHashMap<>();
	PropertiesConfigurationFactory<Map<String, Object>> factory
			= new PropertiesConfigurationFactory<>(content);
	factory.setTargetName("git");
	Properties properties = PropertiesLoaderUtils.loadProperties(resource);
	factory.setProperties(properties);
	factory.bindPropertiesToTarget();
	return new SimpleInfoContributor("gitfull", content);
}
```

I guess we could abstract the binding part.
